### PR TITLE
Handling Spread Function inside Sum Function with dim

### LINKED ISSUE #5546: Handle IntrinsicArrayFunction inside a IntrinsicArrayFunction
Code : 
```fortran
program test
integer :: A(2) = [1,2]
integer :: n = 2
integer :: dim = 1
A = sum(spread(A, dim, n), dim=dim)
print *, n
end program
```

LFortran Error : 
```console
Traceback (most recent call last):
  Binary file "/mnt/c/lfortran/src/bin/lfortran", in _start
  File "/build/glibc-LcI20x/glibc-2.31/csu/../csu/libc-start.c", line 308, in __libc_start_main
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2725, in main
    return main_app(argc, argv);
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2664, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 1027, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    result = fe.get_asr2(input, lm, diagnostics);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 275, in LCompilers::FortranEvaluator::get_asr2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::diag::Diagnostics&)
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 297, in LCompilers::FortranEvaluator::get_asr3(LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&)
    compiler_options.symtab_only, compiler_options);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 101, in LCompilers::LFortran::ast_to_asr(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, bool, LCompilers::CompilerOptions&)
    instantiate_types, instantiate_symbols, entry_functions, entry_function_arguments_mapping, data_structure);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 4618, in LCompilers::LFortran::body_visitor(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::ASR::asr_t*, LCompilers::CompilerOptions&, std::map<unsigned long, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned long, LCompilers::ASR::symbol_t*, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, LCompilers::ASR::symbol_t*> > >&, std::map<unsigned long, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<int, std::allocator<int> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<int, std::allocator<int> > > > >&, std::vector<LCompilers::ASR::stmt_t*, std::allocator<LCompilers::ASR::stmt_t*> >&)
    b.visit_TranslationUnit(ast);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 154, in LCompilers::LFortran::BodyVisitor::visit_TranslationUnit(LCompilers::LFortran::AST::TranslationUnit_t const&)
    visit_ast(*x.m_items[i]);
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4926, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_ast(LCompilers::LFortran::AST::ast_t const&)
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4882, in void LCompilers::LFortran::AST::visit_ast_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::ast_t const&, LCompilers::LFortran::BodyVisitor&)
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4929, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_mod(LCompilers::LFortran::AST::mod_t const&)
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4507, in void LCompilers::LFortran::AST::visit_mod_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::mod_t const&, LCompilers::LFortran::BodyVisitor&)
    case modType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 1927, in LCompilers::LFortran::BodyVisitor::visit_Program(LCompilers::LFortran::AST::Program_t const&)
    transform_stmts(body, x.n_body, x.m_body);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 117, in LCompilers::LFortran::BodyVisitor::transform_stmts(LCompilers::Vec<LCompilers::ASR::stmt_t*>&, unsigned long, LCompilers::LFortran::AST::stmt_t**)
    this->visit_stmt(*m_body[i]);
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4974, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_stmt(LCompilers::LFortran::AST::stmt_t const&)
    void visit_stmt(const stmt_t &b) { visit_stmt_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4617, in void LCompilers::LFortran::AST::visit_stmt_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::stmt_t const&, LCompilers::LFortran::BodyVisitor&)
    case stmtType::Assignment: { v.visit_Assignment((const Assignment_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_body_visitor.cpp", line 2794, in LCompilers::LFortran::BodyVisitor::visit_Assignment(LCompilers::LFortran::AST::Assignment_t const&)
    this->visit_expr(*x.m_value);
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 5026, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::BodyVisitor>::visit_expr(LCompilers::LFortran::AST::expr_t const&)
    void visit_expr(const expr_t &b) { visit_expr_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4680, in void LCompilers::LFortran::AST::visit_expr_t<LCompilers::LFortran::BodyVisitor>(LCompilers::LFortran::AST::expr_t const&, LCompilers::LFortran::BodyVisitor&)
    case exprType::FuncCallOrArray: { v.visit_FuncCallOrArray((const FuncCallOrArray_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 7499, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::visit_FuncCallOrArray(LCompilers::LFortran::AST::FuncCallOrArray_t const&)
    v = intrinsic_as_node(x, is_function);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_common_visitor.h", line 6686, in LCompilers::LFortran::CommonVisitor<LCompilers::LFortran::BodyVisitor>::intrinsic_as_node(LCompilers::LFortran::AST::FuncCallOrArray_t const&, bool&)
    tmp = create_func(al, x.base.base.loc, args, diag);
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_array_function_registry.h", line 3127, in LCompilers::ASRUtils::Sum::create_Sum(Allocator&, LCompilers::Location const&, LCompilers::Vec<LCompilers::ASR::expr_t*>&, LCompilers::diag::Diagnostics&)
    return ArrIntrinsic::create_ArrIntrinsic(al, loc, args, diag,
  File "/mnt/c/lfortran/src/libasr/pass/intrinsic_array_function_registry.h", line 695, in LCompilers::ASRUtils::ArrIntrinsic::create_ArrIntrinsic(Allocator&, LCompilers::Location const&, LCompilers::Vec<LCompilers::ASR::expr_t*>&, LCompilers::diag::Diagnostics&, LCompilers::ASRUtils::IntrinsicArrayFunctions)
    args_merge.push_back(al, b.ArraySize(args[0], b.i32(it+2), int32));
  File "/mnt/c/lfortran/src/libasr/asr_builder.h", line 290, in LCompilers::ASRUtils::ASRBuilder::ArraySize(LCompilers::ASR::expr_t*, LCompilers::ASR::expr_t*, LCompilers::ASR::ttype_t*)
    return EXPR(make_ArraySize_t_util(al, loc, x, dim, t, nullptr));
  File "/mnt/c/lfortran/src/libasr/asr_utils.h", line 80, in LCompilers::ASRUtils::EXPR(LCompilers::ASR::asr_t const*)
    return ASR::down_cast<ASR::expr_t>(f);
  File "/mnt/c/lfortran/src/libasr/asr.h", line 41, in LCompilers::ASR::expr_t* LCompilers::ASR::down_cast<LCompilers::ASR::expr_t, LCompilers::ASR::asr_t>(LCompilers::ASR::asr_t const*)
    LCOMPILERS_ASSERT(is_a<T>(*f));
  File "/mnt/c/lfortran/src/libasr/asr.h", line 32, in bool LCompilers::ASR::is_a<LCompilers::ASR::expr_t, LCompilers::ASR::asr_t>(LCompilers::ASR::asr_t const&)
    return T::class_type == x.type;
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in <invalid>
Segfault: Signal SIGSEGV (segmentation fault) received
```

### PR DESCRIPTION
**Pull request by AnshMehta1**
Fixes #5546 

MRE 
```
program test
integer :: A(2) = [1,2]
integer :: n = 2
integer :: dim = 1
A = sum(spread(A, dim, n), dim=dim)
print *, A
end program
```

LFortran output
```
(lf) lordansh@LordAnsh:~/dev/lfort3/lfortran$ ./src/bin/lfortran try.f90
2    4
```

### CODE REVIEW

**Review by Pranavchiku — APPROVED** [review]
This makes sense, we are giving priority to `array_func_type` and if it is not array pull over array argument. Looks good to me!

File: integration_tests/arrays_65.f90 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+program test
```
Comment (Pranavchiku) [inline integration_tests/arrays_65.f90:1]:
```suggestion
program arrays_45
```

This is minor, do it in subsequent PRs.

**Comment by Pranavchiku** [comment]
Before merging check PRIMA branch `lf15` with this, ensure it does not fail.

**Comment by AnshMehta1** [comment]
> Before merging check PRIMA branch `lf15` with this, ensure it does not fail.

It falis for case where dim (sum argument) is provided with compile - time value . That generates an error due to some other issue.
ASR generates for below MREs.

MRE - 1 :
```
program test
integer :: A(2) = [1,2]
integer :: n = 2
integer :: dim = 1
A = sum(spread(A, dim, n), dim=1)
print *, A
end program
```

Output
```
(lf) lordansh@LordAnsh:~/dev/lfort3/lfortran$ ./src/bin/lfortran try.f90
Segmentation fault (core dumped)
```

MRE - 2 : 
```
program test
integer :: A(2) = [1,2]
integer :: n = 2
integer :: dim = 1
A = sum(spread(A, dim, n), 1)
print *, A
end program
```

Output
```
(lf) lordansh@LordAnsh:~/dev/lfort3/lfortran$ ./src/bin/lfortran try.f90
Segmentation fault (core dumped)
```

File: integration_tests/arrays_65.f90 (hunk 1 above)
Comment (AnshMehta1) [reply integration_tests/arrays_65.f90]: Done

**Comment by Pranavchiku** [comment]
Segmentation fault is fine, PRIMA throws no compilation error, right?

**Comment by Pranavchiku** [comment]
Tested, works well.

**Comment by AnshMehta1** [comment]
> Segmentation fault is fine, PRIMA throws no compilation error, right?

Ya , working well

**Pull request merged**